spotbugs/src/main/java/edu/umd/cs/findbugs/detect/PublicAttributes.java

```diff
@@ -0,0 +1,210 @@
+                return;
+            }
+
+            if (!field.isPublic() || !field.isReferenceType()) {
+                return;
+            }
+
+            Field clsField = getClassField(field, getClassName());
+            Type type = clsField.getType();
+            // Instances of an immutable classes cannot be modified, thus
+            // the method is not a modifier, despite its name.
+            if (type instanceof ObjectType &&
+                    isImmutable(((ObjectType) type).getClassName())) {
+                return;
+            }
+
+            bugReporter.reportBug(new BugInstance(this, "PA_PUBLIC_ATTRIBUTES",
+                    NORMAL_PRIORITY)
+                            .addClass(this).addField(field));
+            writtenFields.add(field);
+        }
+    }
+
+    private static boolean isConstructorLikeMethod(String methodName) {
+        return Const.CONSTRUCTOR_NAME.equals(methodName) ||
+                Const.STATIC_INITIALIZER_NAME.equals(methodName) ||
+                "clone".equals(methodName) ||
+                "init".equals(methodName) ||
+                "initialize".equals(methodName) ||
+                "dispose".equals(methodName) ||
+                "finalize".equals(methodName) ||
+                "this".equals(methodName) ||
+                "_jspInit".equals(methodName) ||
+                "_jspDestroy".equals(methodName);
+    }
+
+    private static Field getClassField(XField field, String className) {
+        try {
+            JavaClass cls = Repository.lookupClass(className);
+            for (Field clsField : cls.getFields()) {
+                if (field.getName().equals(clsField.getName())) {
+                    return clsField;
+                }
+            }
+        } catch (ClassNotFoundException cnfe) {
+            assert false;
+        }
+        return null;
+    }
+
+    private static boolean isImmutable(String className) {
+        if ("java.lang.String".equals(className) ||
+                "java.lang.Character".equals(className) ||
+                "java.lang.Byte".equals(className) ||
+                "java.lang.Short".equals(className) ||
+                "java.lang.Integer".equals(className) ||
+                "java.lang.Long".equals(className) ||
+                "java.lang.Float".equals(className) ||
+                "java.lang.Double".equals(className) ||
+                "java.lang.Boolean".equals(className)) {
+            return true;
+        }
+
+        try {
+            JavaClass cls = Repository.lookupClass(className);
+
+            if (!cls.isFinal()) {
+                return false;
+            }
+
+            for (Field field : cls.getFields()) {
+                if (!field.isStatic() &&
+                        (!field.isFinal() || !field.isPrivate())) {
+                    return false;
+                }
+            }
+
+            return true;
+        } catch (ClassNotFoundException cnfe) {
+            assert false;
```

Comment:
Invoke `AnalysisContext.reportMissingClass(cnfe)` instead, then we can tell the reported problem to end users.